tachyon/zk/plonk/vanishing/data/vanishing_permutation.h

```diff
@@ -0,0 +1,79 @@
+// Copyright 2020-2022 The Electric Coin Company
+// Copyright 2022 The Halo2 developers
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
+// file.
+
+#ifndef TACHYON_ZK_PLONK_VANISHING_DATA_VANISHING_PERMUTATION_H_
+#define TACHYON_ZK_PLONK_VANISHING_DATA_VANISHING_PERMUTATION_H_
+
+#include <utility>
+#include <vector>
+
+#include "absl/types/span.h"
+
+#include "tachyon/zk/plonk/circuit/rotation.h"
+#include "tachyon/zk/plonk/permutation/permutation_committed.h"
+#include "tachyon/zk/plonk/permutation/permutation_proving_key.h"
+#include "tachyon/zk/plonk/vanishing/data/vanishing_common.h"
+#include "tachyon/zk/plonk/vanishing/vanishing_utils.h"
+
+namespace tachyon::zk {
+
+// It stores PERMUTATION_RELATED values when building a polynomial below,
+// gate₀(X) + y * gate₁(X) + ... + yⁱgateᵢ(X) + ...
+// See Halo2 book in detail,
+// https://zcash.github.io/halo2/design/proving-system/vanishing.html
+template <typename Poly, typename Evals>
+class VanishingPermutation {
+ public:
+  using F = typename Evals::Field;
+
+  template <typename PCSTy>
+  static VanishingPermutation Create(
+      ProverBase<PCSTy>* prover, const VanishingCommon<F>* common,
+      const PermutationCommitted<Poly>& committed_permutation,
+      const PermutationProvingKey<Poly, Evals>& permutation_proving_key) {
+    size_t blinding_factors = prover->blinder().blinding_factors();
+    Rotation last_rotation =
+        Rotation(-static_cast<int32_t>(blinding_factors + 1));
```

Comment:
I think it's good to use checked math because it tries to narrow down and do we have a runtime guarantee that this always is a valid value, right?